### PR TITLE
Issue 13955 prevent that user can add contentlet to a content type in a container with no permission

### PR DESCRIPTION
**Pull request by freddyucv**

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPI.java [hunk 1]
```diff
@@ -97,4 +100,25 @@ public boolean doesUserHasPermission (String containerInode, int permission, boo
...
+			if(!InodeUtils.isSet(containerInode)) {
+				return false;
+			} else {
+				final User systemUser = APILocator.getUserAPI().getSystemUser();
+				final Container container = APILocator.getContainerAPI().find(containerInode, systemUser, false);
+				User loggedInUser = APILocator.getLoginServiceAPI().getLoggedInUser();
+				List<ContentType> contentTypesInContainer = containerAPI.getContentTypesInContainer(loggedInUser, container);
+				return !contentTypesInContainer.isEmpty();
+			}
+		} catch (DotSecurityException e) {
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPI.java]: what happens with this exception? At least, it should be logged

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPI.java [hunk 2]
```diff
@@ -97,4 +100,25 @@ public boolean doesUserHasPermission (String containerInode, int permission, boo
 			return false;
 		}
 	}
+
+
+	/**
+	 * This method checks if the logged in user has the required permission to ADD any content into the
+	 * the passed container id
+	 */
+	public boolean doesUserHasPermissionToAddContent (String containerInode) throws DotDataException {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPI.java]: set to final

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPI.java [hunk 3]
```diff
@@ -97,4 +100,25 @@ public boolean doesUserHasPermission (String containerInode, int permission, boo
...
+	 * the passed container id
+	 */
+	public boolean doesUserHasPermissionToAddContent (String containerInode) throws DotDataException {
+		try {
+			if(!InodeUtils.isSet(containerInode)) {
+				return false;
+			} else {
+				final User systemUser = APILocator.getUserAPI().getSystemUser();
+				final Container container = APILocator.getContainerAPI().find(containerInode, systemUser, false);
+				User loggedInUser = APILocator.getLoginServiceAPI().getLoggedInUser();
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPI.java]: set to final

File: dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java [hunk 4]
```diff
@@ -353,18 +347,62 @@ public void saveContainerStructures(final List<ContainerStructure> containerStru
...
+					cTypeList.add(type);
+				} catch (DotSecurityException e) {
+					continue;
+				}
+			}
+
+			return cTypeList.stream().sorted(Comparator.comparing(ContentType::name)).collect(Collectors.toList());
+
+		} catch (DotSecurityException e) {
+			return Collections.EMPTY_LIST;
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java]: can you log the exception?

File: dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java [hunk 5]
```diff
@@ -353,18 +347,62 @@ public void saveContainerStructures(final List<ContainerStructure> containerStru
     @Override
     public List<ContentType> getContentTypesInContainer(Container container) throws DotStateException, DotDataException, DotSecurityException  {
 
-        List<ContainerStructure> csList =  getContainerStructures(container);
-        List<ContentType> cTypeList = new ArrayList<ContentType>();
+		try {
+			List<ContainerStructure>  csList = getContainerStructures(container);
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java]: rename to containerList

File: dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java [hunk 6]
```diff
@@ -353,18 +347,62 @@ public void saveContainerStructures(final List<ContainerStructure> containerStru
...
+					cTypeList.add(type);
+				} catch (DotSecurityException e) {
+					continue;
+				}
+			}
+
+			return cTypeList.stream().sorted(Comparator.comparing(ContentType::name)).collect(Collectors.toList());
+
+		} catch (DotSecurityException e) {
+			return Collections.EMPTY_LIST;
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java]: same here

File: dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java [hunk 7]
```diff
@@ -353,18 +347,62 @@ public void saveContainerStructures(final List<ContainerStructure> containerStru
...
+			List<ContainerStructure>  csList = getContainerStructures(container);
 
-        for (ContainerStructure cs : csList) {
-            ContentType type = APILocator.getContentTypeAPI(APILocator.systemUser()).find(cs.getStructureId());
-            if(type==null) continue;
-            cTypeList.add(type);
-        }
-        
-        return cTypeList.stream().sorted(Comparator.comparing(ContentType::name)).collect(Collectors.toList());
+			List<ContentType> cTypeList = new ArrayList<>();
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java]: rename to contentTypeList

File: dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java [hunk 8]
```diff
@@ -353,18 +347,62 @@ public void saveContainerStructures(final List<ContainerStructure> containerStru
...
+
+			for (ContainerStructure cs : csList) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java]: set to final and rename to containerStructure

File: dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java [hunk 9]
```diff
@@ -353,18 +347,62 @@ public void saveContainerStructures(final List<ContainerStructure> containerStru
...
+				try {
+					ContentType type = APILocator.getContentTypeAPI(APILocator.systemUser()).find(cs.getStructureId());
+					if(type == null) continue;
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java]: avoid using if without curly bracket, must be such as 

if (null == type) {
   continue;
}

File: dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java [hunk 10]
```diff
@@ -353,18 +347,62 @@ public void saveContainerStructures(final List<ContainerStructure> containerStru
...
+					cTypeList.add(type);
+				} catch (DotSecurityException e) {
+					continue;
+				}
+			}
+
+			return cTypeList.stream().sorted(Comparator.comparing(ContentType::name)).collect(Collectors.toList());
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java]: if this should be immutable you can use CollectionUtils.toImmutableList()

File: dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java [hunk 11]
```diff
@@ -353,18 +347,62 @@ public void saveContainerStructures(final List<ContainerStructure> containerStru
...
+
+		} catch (DotSecurityException e) {
+			return Collections.EMPTY_LIST;
+		}
     }
-    
+
+	@CloseDBIfOpened
+	@Override
+	public List<ContentType> getContentTypesInContainer(User user, Container container) throws
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java]: set to final

File: dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java [hunk 12]
```diff
@@ -353,18 +347,62 @@ public void saveContainerStructures(final List<ContainerStructure> containerStru
...
+			DotStateException, DotDataException {
+
+		try {
+			List<ContainerStructure>  csList = getContainerStructures(container);
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java]: rename ti to containerStructureList

File: dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java [hunk 13]
```diff
@@ -353,18 +347,62 @@ public void saveContainerStructures(final List<ContainerStructure> containerStru
...
+
+			List<ContentType> cTypeList = new ArrayList<>();
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java]: rename to contentTypeList

File: dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java [hunk 14]
```diff
@@ -353,18 +347,62 @@ public void saveContainerStructures(final List<ContainerStructure> containerStru
...
+
+			for (ContainerStructure cs : csList) {
+				try {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java]: basically the same comments I made previously

File: dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java (hunk 4 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java]: this exception really never happend, this exception is suppose to be throw by  'getContainerStructures' method according its signature:

https://github.com/dotCMS/core/blob/55da87fb8d5803f97bc0d92ffb90fd12ec83baae/dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java#L325

but this method really never throw that exception

https://github.com/dotCMS/core/blob/55da87fb8d5803f97bc0d92ffb90fd12ec83baae/dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java#L322

Comment (freddyucv) [reply dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java]: Also the method return all the content types that can be add by user into a specific container, so if the user can get the container Object then the method return a empty list

**Review by jdotcms — APPROVED** [review]

**Review by nollymar — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPI.java [hunk 15]
```diff
@@ -97,4 +101,32 @@ public boolean doesUserHasPermission (String containerInode, int permission, boo
 			return false;
 		}
 	}
+
+
+	/**
+	 * This method checks if the logged in user has the required permission to ADD any content into the
+	 * the passed container id
+	 */
+	public boolean doesUserHasPermissionToAddContent (final String containerInode) throws DotDataException {
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPI.java]: Integration test for this new method? @freddyucv

File: dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPI.java [hunk 16]
```diff
@@ -256,4 +256,16 @@ public void deleteContainerContentTypesByContainerInode(final Container containe
...
+	 * Return the {@link ContentType} into a {@link Container} for a specific {@link User}, return a empty List if the
+	 * user don't have READ permission in any {@link ContentType} into the specific {@link Container}
+	 *
+	 * @param user
+	 * @param container
+	 * @return
+	 * @throws DotStateException
+	 * @throws DotDataException
+	 */
+	List<ContentType> getContentTypesInContainer(User user, Container container) throws DotStateException, DotDataException;
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPI.java]: Integration test for this method? @freddyucv

**Review by dsilvam — APPROVED** [review]

**Pull request merged**